### PR TITLE
Grpc

### PR DESCRIPTION
**Pull request by PavelDonchenko**

### CODE REVIEW

File: server/Dockerfile [hunk 1]
```diff
@@ -3,9 +3,9 @@ WORKDIR /app
 COPY go.mod ./
 COPY go.sum ./
 RUN go mod download
-COPY . .
+COPY ../.. .
 RUN go build -o /main
 
-EXPOSE 8800
+
```
Comment (penkova) [inline server/Dockerfile]: why do you remove exposing?

File: client/provider/rest/user/user_handler.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	c *controllers.BaseHandler
+}
+
+func NewRouterHandler(c *controllers.BaseHandler) *routerHandler {
+	return &routerHandler{
+		c: c,
+	}
+}
+
+func (h routerHandler) GetAll(ctx *gin.Context) {
```
Comment (penkova) [inline client/provider/rest/user/user_handler.go]: please rename it to `GetAllUsers`

File: client/provider/rest/user/user_handler.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	res, err := h.c.GetAll()
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	ctx.JSON(http.StatusOK, gin.H{"response": res})
+}
+
+func (h routerHandler) GetOne(ctx *gin.Context) {
```
Comment (penkova) [inline client/provider/rest/user/user_handler.go]: please rename it to `GetUserByID` or `GetUser`

File: client/provider/rest/user/user_handler.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	res, err := h.c.Get(uint32(userId))
+	if err != nil {
+		ctx.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
+		return
+	}
+
+	ctx.JSON(http.StatusOK, gin.H{"response": res})
+}
+
+func (h routerHandler) Create(ctx *gin.Context) {
```
Comment (penkova) [inline client/provider/rest/user/user_handler.go]: rename ...

File: client/provider/rest/user/user_handler.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	res, err := h.c.Create(&user)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, err.Error())
+		return
+	}
+
+	ctx.JSON(http.StatusOK, gin.H{"response": res})
+}
+
+func (h routerHandler) Update(ctx *gin.Context) {
```
Comment (penkova) [inline client/provider/rest/user/user_handler.go]: rename ...

File: client/provider/rest/user/user_handler.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	ctx.JSON(http.StatusOK, gin.H{"response": resUpdate})
+}
+
+func (h routerHandler) Delete(ctx *gin.Context) {
```
Comment (penkova) [inline client/provider/rest/user/user_handler.go]: rename ...

File: client/main.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	client := pb.NewUserClient(conn)
+
+	c := controllers.NewBaseHandler(client)
+
+	rest.CreateRouter(c)
+
```
Comment (penkova) [inline client/main.go:24]: please remove this empty line

File: client/provider/rest/router.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func CreateRouter(c *controllers.BaseHandler) {
+	h := rest.NewRouterHandler(c)
+	router := gin.Default()
+
+	router.GET("/users/:id", h.GetOne)
+	router.GET("/users", h.GetAll)
+	router.POST("/users", h.Create)
+	router.PUT("/users/update", h.Update)
```
Comment (penkova) [inline client/provider/rest/router.go]: don't use verbs in url schema, will be better

```
router.GET("/users/:id", h.GetOne)
	router.GET("/users", h.GetAll)
	router.POST("/users", h.Create)
	router.PUT("/users/:id", h.Update)
	router.DELETE("/users/:id", h.Delete)
```

File: Makefile [hunk 9]
```diff
@@ -1,22 +1,42 @@
...
+
 stop: ## Stop running containers
-	docker stop
+	docker-compose stop
 
 rm: stop ## Stop and remove running containers
 	docker rm $(APP_NAME)
 
-lint: ## Run golangci-lint
+lintS: ## Run golangci-lint on Server
```
Comment (penkova) [inline Makefile]: `lint-server`

File: Makefile [hunk 10]
```diff
@@ -1,22 +1,42 @@
...
+	cd server; \
+	golangci-lint run
+	cd server; \
+	go vet ./...
+	echo "Golangci-lint and vet tests are finished successful"
+
+lintC: ## Run golangci-lint on Client
```
Comment (penkova) [inline Makefile]: `lint-client`

File: pkg/controllers/connect.go [hunk 11]
```diff
@@ -16,10 +16,10 @@ type Server struct {
 	Router *mux.Router
 }
 
-func (s *Server) Initialize(Dbdriver string) {
+func (s *Server) Initialize(Dbdriver, user, password, host, dbname string) {
 	var err error
 	if Dbdriver == "mysql" {
-		dns := "pavel:mysqlpaha100688@tcp(bookstore-mysql:3306)/testdb2?charset=utf8mb4&parseTime=True&loc=Local"
+		dns := user + ":" + password + "@tcp(" + host + ")/" + dbname + "?charset=utf8mb4&parseTime=True&loc=Local"
 		s.DB, err = gorm.Open(Dbdriver, dns)
```
Comment (penkova) [inline pkg/controllers/connect.go]: please move the opening of the DB connection to a separate file on the repository layer

File: server/cmd/server/server.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+		if err != nil {
+			fmt.Printf("Cannot connect to %s database\n", Dbdriver)
+			log.Fatal("This is the error:", err)
+		} else {
+			fmt.Printf("We are connected to the %s database\n", Dbdriver)
+		}
+	}
+
+	return DB
+	//RegisterBookStoreRoutes()
```
Comment (penkova) [inline server/cmd/server/server.go]: do you still need this line?

File: server/proto/user.proto [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+message GetUserRequest {
+    uint32 Id = 1;
+}
+
+message GetUserResponse {
+    UserItem user = 1;
+}
+
+message GetAllAUserRequest {}
```
Comment (penkova) [inline server/proto/user.proto]: how about pagination?

File: server/service/user/service.go [hunk 14]
```diff
@@ -0,0 +1,14 @@
+package service
+
+import (
+	pb "github.com/PavelDonchenko/bookstorejRPC/server/gen/proto"
+	model "github.com/PavelDonchenko/bookstorejRPC/server/models"
+)
+
+type UService interface {
```
Comment (penkova) [inline server/service/user/service.go]: UService is bad name

File: server/service/user/user.go [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+	items := []*pb.UserItem{}
+
+	for _, v := range users {
+		u := &pb.UserItem{
+			Id:       uint32(v.ID),
+			Nickname: v.Nickname,
+			Email:    v.Email,
+			Password: v.Password,
+			CrateAt:  timestamppb.Now(),
```
Comment (penkova) [inline server/service/user/user.go]: please update it according to our discussion

**Comment by penkova** [comment]
Please add readme file with instruction how run your project and describe main commands

File: Makefile [hunk 16]
```diff
@@ -1,22 +1,42 @@
...
 
+create_proto:
+	cd server; \
+	protoc --proto_path=proto proto/*.proto --go_out=gen/
+	cd server; \
+	protoc --proto_path=proto proto/*.proto --go-grpc_out=gen/
+
+clean_proto:
+	rm gen/proto/*.go
 build: ## Build the release and develoment container. The development
```
Comment (penkova) [inline Makefile:21]: add new line before build command

File: server/Dockerfile (hunk 1 above)
Comment (PavelDonchenko) [reply server/Dockerfile]: +

Comment (PavelDonchenko) [reply server/Dockerfile]: fixed

File: client/provider/rest/user/user_handler.go (hunk 2 above)
Comment (PavelDonchenko) [reply client/provider/rest/user/user_handler.go]: fixed

File: client/provider/rest/user/user_handler.go (hunk 3 above)
Comment (PavelDonchenko) [reply client/provider/rest/user/user_handler.go]: fixed

File: client/provider/rest/user/user_handler.go (hunk 4 above)
Comment (PavelDonchenko) [reply client/provider/rest/user/user_handler.go]: fixed

File: client/provider/rest/user/user_handler.go (hunk 5 above)
Comment (PavelDonchenko) [reply client/provider/rest/user/user_handler.go]: fixed

File: client/provider/rest/user/user_handler.go (hunk 6 above)
Comment (PavelDonchenko) [reply client/provider/rest/user/user_handler.go]: fixed

File: client/main.go (hunk 7 above)
Comment (PavelDonchenko) [reply client/main.go:24]: fixed

File: Makefile (hunk 9 above)
Comment (PavelDonchenko) [reply Makefile]: fixed

File: Makefile (hunk 10 above)
Comment (PavelDonchenko) [reply Makefile]: fixed

File: pkg/controllers/connect.go (hunk 11 above)
Comment (PavelDonchenko) [reply pkg/controllers/connect.go]: fixed

File: server/cmd/server/server.go (hunk 12 above)
Comment (PavelDonchenko) [reply server/cmd/server/server.go]: deleted

File: server/service/user/user.go (hunk 15 above)
Comment (PavelDonchenko) [reply server/service/user/user.go]: fixed

File: Makefile (hunk 16 above)
Comment (PavelDonchenko) [reply Makefile:21]: fixed

File: server/proto/user.proto (hunk 13 above)
Comment (PavelDonchenko) [reply server/proto/user.proto]: Implemented

File: server/service/user/service.go (hunk 14 above)
Comment (PavelDonchenko) [reply server/service/user/service.go]: renamed

**Review by penkova — APPROVED** [review]

**Pull request merged**